### DISCUSSION TITLE
Where are skills assigned to tasks?

### DISCUSSION THREAD

**Discussion by peyre** [raceintospace/raceintospace#596] [Q&A] [answered]
I see that the names of mission steps are defined in game_main.cpp.  A number of times I've looked to see where they are assigned skills to apply to the hardware's Safety factor for a given step, such as Reentry being a Capsule step.  Anyone know where that happens?

**Comment by hweimer** [comment, marked as answer]
This is done in MissionSteps() in mc2.cpp. https://github.com/raceintospace/raceintospace/blob/a3ecec0d123793b561652e29478eebc438a66537/src/game/mc2.cpp#L202

**Reply by peyre** [reply]
Ohh, thanks.  So mcode identifies the mission index code I think, and Mev[step] would give the step number for that mission.  What I don't understand is why the mission code seems to be given as (usually) a letter.  Does A, B, C = 1, 2, 3; a, b, c = 27, 28, 29, etc?

**Reply by peyre** [reply]
I would assume the steps are defined in misSteps.dat, but they show as things like "A1 B0 F1000 M10 N F0011 D00 F0000 G0111", which I'm not sure what to make of.  If I could understand these, I could add comments to the file explaining what they all are, for future reference.

**Reply by hweimer** [reply]
mcode is a letter because it is derived from the mission string in mission.dat, which is in a semi-human-readable form. You are correct about the translation rule into numbers, except that 'A' gets mapped to 0 and hence all the numbers are shifted down by one. The number is then stored in Mev[step].loc. If you need to know, which step a particular number (or letter) corresponds to, you can look at the array S_Name in game_main.cpp.

The numbers in missSteps.dat are only for displaying the mission trajectory in the future mission screen, they are not used for the actual missions. That information comes from the mission string, which also contains the steps for alternate branches.

**Reply by peyre** [reply]
Thanks!

**Comment by peyre** [comment]
Thanks Hendrik.  Getting back to this, comparing the two files is confusing.  Suborbital, for instance, seems to be steps ADE, yet mc2.cpp (under void MissionSteps or MissionParse) doesn't seem to define either D or E.  Similarly, Earth Orbital looks to be steps ABFGDE, but none of those seem to be defined except A.  Clearly I'm missing something here...

    On Saturday, October 30, 2021, 12:16:33 PM PDT, Hendrik W. ***@***.***> wrote:  
 
 


mcode is a letter because it is derived from the mission string in mission.dat, which is in a semi-human-readable form. You are correct about the translation rule into numbers, except that 'A' gets mapped to 0 and hence all the numbers are shifted down by one. The number is then stored in Mev[step].loc. If you need to know, which step a particular number (or letter) corresponds to, you can look at the array S_Name in game_main.cpp.

The numbers in missSteps.dat are only for displaying the mission trajectory in the future mission screen, they are not used for the actual missions. That information comes from the mission string, which also contains the steps for alternate branches.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub, or unsubscribe.
Triage notifications on the go with GitHub Mobile for iOS or Android.

**Reply by hweimer** [reply]
Not sure what you mean by "define" in this context. The relevant line is this one:
https://github.com/raceintospace/raceintospace/blob/e1917d91aae4669817c7cac7b13dd848ea92132d/src/game/mc2.cpp#L618
mcode is the letter of the step and 65 is 'A' in ASCII. So, if you have a 'D', Mev[step].loc becomes 3, which is the internal code for a re-entry step and controls which videos are shown, which failures are generated, etc.

Does this address your concern?

**Comment by peyre** [comment]
Ok, I get that A=65 in ASCII.  What I'm getting at with all this is, if we were to implement the "Docking and Navigation" that Fritz had in mind (as in #384), how would we change certain steps from relying on Capsule skill to instead rely on Docking skill?  I'm afraid this looks way more complicated than I had imagined.

https://github.com/raceintospace/raceintospace/issues/384
    On Tuesday, October 11, 2022 at 08:15:20 AM PDT, Hendrik W. ***@***.***> wrote:  
 
 


Not sure what you mean by "define" in this context. The relevant line is this one:
https://github.com/raceintospace/raceintospace/blob/e1917d91aae4669817c7cac7b13dd848ea92132d/src/game/mc2.cpp#L618
mcode is the letter of the step and 65 is 'A' in ASCII. So, if you have a 'D', Mev[step].loc becomes 3, which is the internal code for a re-entry step and controls which videos are shown, which failures are generated, etc.

Does this address your concern?

**Reply by hweimer** [reply]
Ah, I see! I think you would have to do it similar to how duration steps are handled. First, you have to introduce a new value for the step class which is set whenever you encounter one of the steps you want to use the docking and navigation skill. Then, whenever a step class has the value corresponding to the new value, you replace the class by the actual hardware and set the .ast variable to DOC[pad] and the .asf variable to MA[pad][Mev[step].ast].A->Docking (like it is done for docking steps). Note that the use of kickers might make things slightly more complicated, depending on what steps you are planning to modify.

**Reply by peyre** [reply]
Ohh, complicated.

**Comment by peyre** [comment]
Of course, if we were to implement this, we'd need to decide which steps to use Docking/Navigation rather than Capsule.  Here's a list from game_main.cpp, with my first guess at which to use in boldface.  I'm sure this needs some revision, but it's a start.

LAUNCH
ORBITAL INS. BURN
HARDWARE POWER-ON
REENTRY
RECOVERY/LANDING
EARTH ORBITAL ACTIVITIES
**EARTH DE-ORBIT BURN**
ORBITAL EVA
**DOCKING** _(already is, of course)_
TRANS-LUNAR INJECTION
TRANS-EARTH INJECTION
**LUNAR ORBITAL INS. BURN**
LUNAR ORBITAL ACTIVITIES
**EARTH MID-COURSE ACT.**
**LUNAR MID-COURSE ACT.**
LUNAR EVA
LEM ACTIVITIES
LUNAR DE-ORBIT BURN
LUNAR LANDING
LUNAR LIFTOFF
PHOTO RECONNAISSANCE
PLANET ORBIT BURN
PLANETARY ACTIVITIES
EARTH RENDEZVOUS
LUNAR RENDEZVOUS
TRANS-PLANETARY INJ.
LEM THRUST TEST
DURATION
JOINT DURATION
**REENTRY CORRIDOR**
**MID-COURSE CORR. BURN**
**MID-COURSE CORR. BURN**
**EARTH ORBITAL INS. BURN**
